### PR TITLE
Feat/sma 809 potential account factories

### PR DESCRIPTION
**Pull request by livingrockrises**
couple of things may be reverted from this change. first, here is the list of changes below and then I would dive into discussion and link the PR for further experimental changes.

**1. Stakeable**

Added common/Stakeable contract (which is Ownable) that can be inherited by the factory [ Or any entity which needs a stake on the entry point for that matter ]

**2. AccountFactory minor changes**

base StakeManager was wrong.
now Factory inherits from Stakeable which makes it also Owned. Using this owner we could further add access controlled actions like whitelisting one or more modules, Or setting ERC7484 registry which checks on module being installed

also it felt right to rename getCounterFactualAddress -> computeAccountAddress :))

**3.small) Took a snapshot of current AccountFactory in case we start editing this one in current branch**
that would mean altering createAccount method from
`function createAccount(address validationModule, bytes calldata moduleInstallData, uint256 index)`
to something 
which can
i. accept generic initData to be directly posted on the account (account.call(initData)) / initData to call specific method on the account INexus.initializeAccount(initData)

ii. accepts array of modules and setup datas, or packed data and factory knows how to decode those and check agaisnt some whitelist.

**4.  BiconomyMetaFactory. inspired from kernel's factory staker**
Stakeable Ownable Factory, idea is to use this as meta factory and it's address will be part of the initcode and under the hood you can use any approved (only approved ones) factory using deployUsingFactory() method.
clear benefit is you can gatekeep factories
if the approved/whitelisted factory has generic call data then issue reported by holterhaus still remains. perhaps these factories can have initial validator/few modules enshrined or whitelisted
in the future you can just add a new factory under the hood and start using it. 

con might be added gas but that comes with any kind of whitelisting anywhere or when addressing security concerns.

**5. Bootstrap.sol**
singleton bootstrap contract inspired from ERC7579 reference implementation 
this contract has util methods to prepare the module installation calldata 

it would work like this in Nexus initialisation

```
function initializeAccount(bytes calldata initData) external payable virtual {
        // checks if already initialized and reverts before setting the state to initialized
        _initModuleManager();

        // this is just implemented for demonstration purposes. You can use any other initialization
        // logic here.
        (address bootstrap, bytes memory bootstrapCall) = abi.decode(initData, (address, bytes));
        (bool success,) = bootstrap.delegatecall(bootstrapCall);
        if (!success) revert();
 }
```
using Bootstrap util contract one can make bootstrap config using array of modules to be installed upon. 
implementation is not altered in this branch to use this bootstrap contract

but experimental changes are here with evolved Bootstrap contract
https://github.com/bcnmy/nexus/blob/test/showcase-alterations-using-bootstrap/contracts/utils/Bootstrap.sol#L82 

P.S - This does not help resolve Holterhaus reported issue but only to allow installing multiple modules. If we go with this approach I haven't thought about how to decode bootstrap config and make checks for individual module addresses or payloads.

### CODE REVIEW

**Comment by livingrockrises** [comment]
also linking this here
https://github.com/bcnmy/nexus/pull/74

File: contracts/utils/Bootstrap.sol [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        _installValidator(address(validator), data);
+    }
+
+    /**
+     * This function is intended to be called by the MSA with a delegatecall.
+     * Make sure that the MSA already initilazed the linked lists in the ModuleManager prior to
+     * calling this function
+     */
+    function initMSA(
+        BootstrapConfig[] calldata $valdiators,
```
Comment (Aboudjem) [inline contracts/utils/Bootstrap.sol:35]: typo

File: contracts/utils/Bootstrap.sol [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+contract Bootstrap is ModuleManager {
+    function singleInitMSA(IModule validator, bytes calldata data) external {
+        // init validator
+        _installValidator(address(validator), data);
+    }
+
+    /**
+     * This function is intended to be called by the MSA with a delegatecall.
+     * Make sure that the MSA already initilazed the linked lists in the ModuleManager prior to
```
Comment (Aboudjem) [inline contracts/utils/Bootstrap.sol:31]: typo

File: contracts/common/Stakeable.sol [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+//
+// ──────────────────────────────────────────────────────────────────────────────
+// Nexus: A suite of contracts for Modular Smart Account compliant with ERC-7579 and ERC-4337, developed by Biconomy.
+// Learn more at https://biconomy.io. For security issues, contact: security@biconomy.io
+
+import {Ownable} from "solady/src/auth/Ownable.sol";
+import {IEntryPoint} from "account-abstraction/contracts/interfaces/IEntryPoint.sol";
+import {IStakeable} from "../interfaces/common/IStakeable.sol";
+
+/**
```
Comment (Aboudjem) [inline contracts/common/Stakeable.sol:19]: can we use `///` there to keep consistency accross all the files

File: contracts/factory/BiconomyMetaFactory.sol [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+    error FactoryNotWhotelisted();
+
+    mapping(address => bool) public factoryWhitelist;
+
+    constructor(address owner) Stakeable(owner) {
+    }
+
+    // could be IAccountFactory
+    // whitelist / blacklist a factory
```
Comment (Aboudjem) [inline contracts/factory/BiconomyMetaFactory.sol:31]: add natspec

File: contracts/factory/BiconomyMetaFactory.sol [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    function whitelistFactory(address factory, bool whitelisted) external payable onlyOwner {
+        factoryWhitelist[factory] = whitelisted;
+    }
+
+    // Note: deploy using only one of the whitelisted factories
```
Comment (Aboudjem) [inline contracts/factory/BiconomyMetaFactory.sol:36]: same here

File: contracts/interfaces/common/IStakeable.sol [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+//    / | / /__ | |/ /_  _______
+//   /  |/ / _ \|   / / / / ___/
+//  / /|  /  __/   / /_/ (__  )
+// /_/ |_/\___/_/|_\__,_/____/
+//
+// ──────────────────────────────────────────────────────────────────────────────
+// Nexus: A suite of contracts for Modular Smart Account compliant with ERC-7579 and ERC-4337, developed by Biconomy.
+// Learn more at https://biconomy.io. For security issues, contact: security@biconomy.io
+
+/**
```
Comment (Aboudjem) [inline contracts/interfaces/common/IStakeable.sol:15]: ///

**Review by Aboudjem — CHANGES_REQUESTED** [review]
.

**Comment by filmakarov** [comment]
I agree to all the proposed changes. 
Ownable can help us managing stuff that will allow to mitigate HolterHus issue
having generic `bytes data` can also be useful. enabling several modules would be really useful

Question re MetaFactory
> it's address will be part of the initcode
can you elaborate on this please?

**Comment by livingrockrises** [comment]
> I agree to all the proposed changes. Ownable can help us managing stuff that will allow to mitigate HolterHus issue having generic `bytes data` can also be useful. enabling several modules would be really useful
> 
> Question re MetaFactory
> 
> > it's address will be part of the initcode
> > can you elaborate on this please?

I made several functions in bootstrap contract for packing things to generic data
one is broad where you can enable n validators, n executors, hook and n fallbacks
I'm trying to think what would be used more often during deployment and what would be done post deployment by choice and less frequently.
so then I thought maybe n validators and 1 hook is enough? preparing data for this scoped init. wdyt?

> address will be part of the initcode
oh i mean this will be the actual factory recognised by initcode userop
initcode = address(meta factory) + selector(deployWithFactory)initData

but doing this both is possible to solve Holterhaus reported issue?
If I make generic data can you show how we would decode and check whitelist?

**Comment by livingrockrises** [comment]
> > i mean this will be the actual factory recognised by initcode userop
> 
> I see. yeah, this is interesting. so this metafactory will be cosidered a factory by 4337. so it has to be staked.
> 
> > but doing this both is possible to solve Holterhaus reported issue? If I make generic data can you show how we would decode and check whitelist?
> 
> Can we move decoding and checking the wl/attestations to the factories that actually deploy the sa?

1. yes, it is stakeable. child factories can be just ownable

2. yes > decoding and checking the wl/attestations to the factories that actually deploy the sa

on this we have to decide which all kinds of factories we need,
a. one with K1 valdiator enshrined?
b. one with x modules just whitelisted. we will need to send large amount of data and decoding logic for these modules.
c. one with registry check on all modules

have to decide on wl vs registry
on previous call w rhinestone attestation flow is not super clear to me so I would refrain from that right away. 

wdyt?
what makes clear sense is meta factory and freedom to intro more factories and whitelist / enshrine / guard trail stuff inside them.
as long as format is clear say createAccount(bytes calldata initData, bytes32 salt)
and inside you're calling account.call(initData) vs perhaps account.initialise(initData) / account.initialise(_validators, _executors, _hook, _fallbacks)  

it would increase gas ofc but have to be reasonable in choices a bit.

**Comment by filmakarov** [comment]
> wdyt? 

I think, with this meta factory concept, we can start with k1-factory and whitelist-factory. And wait until module registry flow is clearer.

**Comment by livingrockrises** [comment]
> > wdyt?
> 
> I think, with this meta factory concept, we can start with k1-factory and whitelist-factory. And wait until module registry flow is clearer.

cool sounds good.
I will start making some changes

**Comment by livingrockrises** [comment]
> > wdyt?
> 
> I think, with this meta factory concept, we can start with k1-factory and whitelist-factory. And wait until module registry flow is clearer.

in this case should I keep methods for all factories same createAccount(_data, _salt) or doing something like factory.call(packedwithfunctionselector) in meta factory?

**Comment by Aboudjem** [comment]
Got some thoughts on the recent PR. I think we can simplify things and make it more efficient


1. Move the init logic into the NexusAccount directly this way we can get rid of `Bootstrap.sol` and `BootstrapUtil.sol`
2. enshrine whitelist of modules in the factory itself or maybe in the Meta (I'm not sure where will it be used here yet as it's used nowhere for the moment)
3. simplifying the initData structure by encoding everything within the NexusAccount reduces the need for decoding and number of  external calls

this will reduce gas with less external calls, simplify initData structure and reduce the number of contracts (files)

```
User
 |
 | Initiates account creation
 v
+----------------------+
| NexusMetaFactory     |
| - Verify Factory     |
+----------------------+
           |
           | deployWithFactory(factory, factoryData)
           v
+----------------------+
| Specific Factory     |
| - Verify Modules     |
| - Initialize Account |
+----------------------+
           |
           | createAccount(initData, salt)
           v
+----------------------+
| NexusAccount         |
| - Initialize Modules |
| - Configure Account  |
+----------------------+
           |
           | Initialization Complete
           v
         Account Created

```

Lmk what your thoughts

**Comment by livingrockrises** [comment]
There are some review comments I am keeping and leaving a note here.

NexusAccountFactory
// Review may not need stakeable here


ModuleWhitelistFactory
// Review if we should verify calldata[0:4] against the function selector of initNexus

K1ValidatorFactory
// Review: if salt should include K1 Validator address as well

File: contracts/utils/Bootstrap.sol [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+        // init fallback
+        for (uint256 i; i < _fallbacks.length; i++) {
+            if (_fallbacks[i].module == address(0)) continue;
+            _installFallbackHandler(_fallbacks[i].module, _fallbacks[i].data);
+        }
+    }
+
+    function _getInitMSACalldata(
+        BootstrapConfig[] calldata $valdiators,
```
[the remark this answers is not in the document]
Comment (livingrockrises) [reply contracts/utils/Bootstrap.sol]: which typo?

**Review by Aboudjem — APPROVED** [review]

**Pull request merged**